### DISCUSSION TITLE
A real LDAP integration?

### DISCUSSION THREAD

**Discussion by helmut72** [zitadel/zitadel#6007] [Ideas]
Is it planned to integrate LDAP in the backend? This means Zitadel reads user information and his group membership that exist in the LDAP server, checks the password against a LDAP server and combine user information from LDAP with Zitadels additional fields, that doesn't exist in LDAP.

For now, LDAP is just an external identity provider. Users see an additional "LDAP" button:
![](https://zitadel.com/docs/assets/images/zitadel_login_ldap-a61b0c766fa32a769c3248641c7b58fa.png)

Thank you.

**Comment by fforootd** [comment]
When the user selects the LDAP button we call the LDAP backend to fetch data and check the password.
If I understand you correctly you want to:

1. Change the UI flow so that it does not appear as an IdP but rather the login form, right? (this will soon be possible with our new login UI)
2. You want to read additional data of the user and store some of it on the identity in ZITADEL. (I think we should extend actions to allow flexibility here)

Is this what you are looking for or do you want to have a scheduled sync?

**Comment by helmut72** [comment]
Yes, like you explained in 1. LDAP integration should be like in Keycloak:
https://www.keycloak.org/docs/latest/server_admin/#_ldap

Or Authelia:
https://www.authelia.com/configuration/first-factor/ldap/

Or Authentik:
https://goauthentik.io/integrations/sources/ldap/

Regarding number 2, if LDAP doesn't provide fields, missing fields should be merged from fields, that Zitadel provides.

**Reply by ISulimanov** [reply]
@fforootd  Hello!
What do you think about @helmut72 proposal? 

I'm trying to figure out the principal difference on approaches LDAP sync VS LDAP IdP. 
@helmut72 in which cases LDAP sync wins over LDAP IdP? Could you please share some user usecases?

**Reply by fforootd** [reply]
I think there is room to extend our LDAP capabilities, even though we did not yet plan things there we would certainly accept PRs at this stage.

**Reply by helmut72** [reply]
> in which cases LDAP sync wins over LDAP IdP? Could you please share some user usecases?

Reason 1 user acceptance:

First you have a login form from the app, because it also supports local users for special cases. For example:
https://gitea.com/user/login

Username
Password
Login with OIDC

User will be instructed to press "Login with...". It's ok for the user pressing one time "login with...". He will be redirected to the IDP and see:
Username
Password
Login with LDAP

Again a "username/password/login with" form and not just username/password. I understand every casual user who think this is ridiculous, not to type username + password into the fields, but pressing again "login with...".

Reason 2 central user management:

Nearly every serious company use Active Directory/LDAP for the user management. Same for me in my private lab. IDP is just an addon to have OIDC and SAML feature, but it's not the central user management ui.

Reason 3 as I wrote above, other IDP supports LDAP as a user source

**Comment by mbosecke** [comment]
I'll second this proposal as it's currently a blocker for us adopting Zitadel. I'll share our perspective and use case.

From our perspective, using LDAP as a source of user information is an _internal_ detail and not one that needs to be exposed to the end-users. We use LDAP internally for employees but want a standalone identity provider for external users to the web apps that we build and don't want to mix the two sources. We are looking for a SSO portal that both internal employees and external users can use without either of them knowing where the _source_ of their data is (i.e. same UX for both). We would hope that the SSO allows both categories of users to reset their password (with some basic attribute synchronization back to the LDAP source for internal employees).

**Comment by el-j** [comment]
is there any chance the ldap idp will work? for now it just cannot find users or finds to many... 
without ldap working zitadel is not a solution. any update when this will work?
<img width="548" alt="image" src="https://github.com/user-attachments/assets/b73abaa6-8052-446d-bcae-3229e6a2a0c3">

background: have a working ldapsearch string that puts out the correct "dn:  cn=name surname,ou=orga,dc=domain,dc=com"
and it is just one. so why does the idp of zitadel is not working with this settings?

**Reply by fforootd** [reply]
Hey, can you share us some context?

Is this a self-hosted zitadel, what version?
What kind of ldap server are you using? Please share the config for ldap as well.

LDAP generally works so it must be some kind of a problem here 😅

**Reply by el-j** [reply]
@fforootd i got it working now. the trick was to add the domain to the bindDN like :
```DOMAINNAME\username```

but i am still wondering: this method now copies the user informations to the zitadel postgresDB. I checked the sync checkbox. now i am wondering. if a user get deactivated or removed or changed in the LDAP is this synced with the zitadel users?
and when i change informations in this user in zitadel, is this synced back to the LDAP user? 

i think i miss read the original question. from my understanding a "real LDAP" would not create a new user-database on zitadel-site...

**Reply by el-j** [reply]
@fforootd it's a compose-stack hosted zitadel with a traefik-proxy and a step-ca for ssl.

**Reply by fforootd** [reply]
Great to know that it now works.

On your question, zitadel makes a copy of the user so that things like other idps and authentication factors can be added to the user. If the user is being disabled in the LDAP it will be then blocked from authenticating in zitadel, so the only thing not yet covered is to disable a session of a user when he reauthenticates. But we could extend zitadel that i checks on reauthentication if the user is still valid in the LDAP.

**Comment by luik8004** [comment]
@el-j could you please share an example of your LDAP configuration. I am trying to add active directory ldap with this configuration
Servers                ldap://test.local:389
BaseDn*                ou=AC,ou=All Departmens,dc=test,dc=local
BindDn*                cn=ldap,ou=AC,ou=All Departmens,dc=test,dc=local
Userbase*                ou=AC,ou=All Departmens,dc=test,dc=local
User filters*            (objectClass=)
User Object Classes    person
ID attribute*            sAMAccountName

but when authorizing in zitadel I get an error 
LDAP Result Code 201 "Filter Compile Error": ldap: finished compiling filter with extra at end: dap))
